keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+
+#### Transitions of the ResizingPod PodCondition
+
+The following diagram shows possible transitions of ResizingPod.Status and
+ResizingPod.Reason fields respectively.
+
+```text
+
+   +-------------------------------------+
+   |                                     |
+   |                                    2|
+   |                                +----v----+
+   |                                |         |
+   |          +---------------------+  false  +-------------------------+
+   |          |                     | Success |                         |
+   |          |                     |         |                         |
+   |          |                     +----+----+                         |
+   |          |                          |                              |
+   |         1|                         3|                             5|
+   |   +------v-----+             +------v-----+             +----------v---------+
+   |   |            |             |            |             |                    |
+   |   |    true    |4            |    true    |7            |        false       |
+   +---+ InProgress <-------------+ PreEmpting <-------------+ FailedNodeCapacity |
+       |            |             |            |             |                    |
+       +------^-----+             +------------+             +----------+---------+
+             6|                                                         |
+              |                                                         |
+              +---------------------------------------------------------+
+
+```
+
+1. Kubelet, on initiating in-place resize.
+1. Kubelet, on successful completion of in-place resize.
+1. Kubelet, on initiating pre-emption of lower priority Pods.
+1. Kubelet, on completing pre-emption of lower priority Pods.
+1. Kubelet, on Node not having capacity to resize Pod.
+1. Kubelet, on initiating in-place resize retry.
+1. Kubelet, on initiating in-place resize retry with pre-emption.
+
+#### Container resource limit update ordering
+
+When in-place resize is desired for multiple Containers in a Pod, Kubelet updates
+resource limit for the Pod and its Containers in the following manner:
+  1. If resource resizing results in net-increase of a resource type (CPU or Memory),
+     Kubelet first updates Pod-level cgroup limit for the resource type, and then
+     updates the Container resource limit.
+  1. If resource resizing results in net-decrease of a resource type, Kubelet first
+     updates the Container resource limit, and then updates Pod-level cgroup limit.
+  1. If resource update results in no net change of a resource type, only the Container
+     resource limits are updated.
+
+In all the above cases, Kubelet applies Container resource limit decreases before
+applying limit increases.
+
+#### Container resource limit update failure handling
+
+For simplicity, if Container resource limits update fails, Kubelet restarts the
```

Comment:
i think this is surprising and goes against the user api wish to not restart the container when making this request.  it seems preferrable to me to not restart the container if a limit is not achieved, but induce reclaim by converging to the desired limit over a period.

Reply:
Yes, this doesn't seem to align with the rest of kubelet's logic, and also goes against the `ResizePolicy`.

Reply:
I took this out of the design, we keep retrying - it is a common K8s pattern.

Reply:
Updated design resolves this issue.